browser/src/main/java/org/netbeans/html/presenters/browser/SimpleServer.java

```diff
@@ -0,0 +1,706 @@
+                try {
+                    s.socket().bind(address);
+                } catch (IOException ex) {
+                    LOG.log(Level.FINE, "Cannot bind to " + at, ex);
+                    continue;
+                }
+                server = s;
+                break;
+            }
+
+            server.register(this.connection, SelectionKey.OP_ACCEPT);
+        }
+        return server;
+    }
+
+    final class Context implements ThreadFactory {
+
+        private final String id;
+        Executor RUN;
+        Thread RUNNER;
+
+        Context(String id) {
+            this.id = id;
+        }
+
+        @Override
+        public Thread newThread(Runnable r) {
+            Thread t = new Thread(r, "Processor for " + id);
+            RUNNER = t;
+            return t;
+        }
+    }
+
+    @Override
+    Context initializeRunner(String id) {
+        Context c = new Context(id);
+        c.RUN = Executors.newSingleThreadExecutor(c);
+        return c;
+    }
+
+    @Override
+    void runSafe(Context c, Runnable r, Fn.Presenter presenter) {
+        class Wrap implements Runnable {
+
+            @Override
+            public void run() {
+                if (presenter != null) {
+                    try ( Closeable c = Fn.activate(presenter)) {
+                        r.run();
+                    } catch (IOException ex) {
+                        // go on
+                    }
+                } else {
+                    r.run();
+                }
+            }
+        }
+        if (c.RUNNER == Thread.currentThread()) {
+            if (presenter != null) {
+                Runnable w = new Wrap();
+                w.run();
+            } else {
+                r.run();
+            }
+        } else {
+            Runnable w = new Wrap();
+            c.RUN.execute(w);
+        }
+    }
+
+    final class ReadHeader {
+
+        private final StringBuilder buffer = new StringBuilder();
+
+        final ReqRes process(SelectionKey key, ByteBuffer chunk) {
+            String text = new String(chunk.array(), 0, chunk.limit(), StandardCharsets.US_ASCII);
+            buffer.append(text);
+            int fullHeader = buffer.indexOf("\r\n\r\n");
+            if (fullHeader == -1) {
+                return null;
+            }
+            String header = text.substring(0, fullHeader);
+
+            Matcher m = PATTERN_GET.matcher(header);
+            String url = m.find() ? m.group(2) : null;
+            String args = url != null && m.groupCount() == 3 ? m.group(3) : null;
+            String method = m.group(1);
+
+            Map<String, String> context;
+            if (args != null) {
+                Map<String, String> c = new HashMap<>();
+                parseArgs(c, args);
+                context = Collections.unmodifiableMap(c);
+            } else {
+                context = Collections.emptyMap();
+            }
+
+            Matcher length = PATTERN_LENGTH.matcher(header);
```

Comment:
Is it ensured, that there can no be chunked transfer happening?

Reply:
Yes, there is 
```
  int fullHeader = buffer.indexOf("\r\n\r\n");
  if (fullHeader == -1) {
    return null;
  }
```
e.g. no processing happens until the end of header is received.